### PR TITLE
Submitting Name Days Goodie

### PR DESCRIPTION
**Pull request by W25**
**What does your Instant Answer do?**
Displays name days for a given name or date. See https://en.wikipedia.org/wiki/Name_day

**What problem does your Instant Answer solve (Why is it better than organic links)?**
The answer is shown instantly, without the need to peruse a website.

**What is the data source for your Instant Answer? (Provide a link if possible)**
https://en.wikipedia.org/wiki/Name_days_in_Poland

**Why did you choose this data source?**
It's available under an open-source license.

**Are there any other alternative (better) data sources?**
The original Instant Answer Idea mentions http://www.namedaycalendar.com/

**What are some example queries that trigger this Instant Answer?**
name day Maria
name days today
1 June name day

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
Polish users.

**Is this Instant Answer connected to a DuckDuckHack [Instant Answer idea](https://duck.co/ideas)?**
https://duck.co/ideas/idea/4590/name-days

**Which existing Instant Answers will this one supercede/overlap with?**
None.

**Are you having any problems? Do you need our help with anything?**
None.

**Where did you hear about DuckDuckHack? (For first time contributors)**
From the DuckDuckGo main page.

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**
nameday

## Checklist
Please place an 'X' where appropriate.

```
[X] Added metadata and attribution information (https://duck.co/duckduckhack/metadata)
[X] Wrote test file and added to t/ directory (https://duck.co/duckduckhack/goodie_tests)
[X] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
[X] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
[] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Windows 8
        [] Google Chrome
        [X] Firefox
        [] Opera
        [] IE 10

    - Windows 7
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 8
        [] IE 9
        [] IE 10

    - Windows XP
        [] IE 7
        [] IE 8

    - Mac OSX
        [] Google Chrome
        [] Firefox
        [] Opera
        [] Safari

    - iOS (iPhone)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - iOS (iPad)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - Android
        [] Firefox
        [] Native Browser
        [] Google Chrome
        [] Opera
```

### CODE REVIEW

**Comment by abeyang** [comment]
Hi @W25, here's what we came up with:

name-days-v2 1-maria

LMK if you have any questions!

**Comment by W25** [comment]
@abeyang Thank you very much, it looks great. Could you please give me the CSS and an example of the HTML code that the goodie should output?

**Comment by W25** [comment]
@abeyang I had tried to implement your design, please take a look.

**Comment by abeyang** [comment]
Great, thanks @W25! Will take a look as soon as I can! /cc @jagtalon

**Comment by jagtalon** [comment]
Thanks @W25! Putting it up on a server. :)

File: t/NameDays.t [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    'imieniny 9.1' => test_zci($jan_9, html => $jan_9_html),
+    'imieniny 9 stycznia' => test_zci($jan_9, html => $jan_9_html),
+    'imieniny 9 styczeń' => test_zci($jan_9, html => $jan_9_html),
+    'imieniny 31 Grudnia' => test_zci($dec_31, html => $dec_31_html),
+    'Imieniny 31 GRUDNIA' => test_zci($dec_31, html => $dec_31_html),
+    'Imieniny 29 lutego' => test_zci($feb_29, html => $feb_29_html),
+    
+    # Czech date formats
+    'svátek 9 ledna' => test_zci($jan_9, html => $jan_9_html),
+    'jmeniny 9 leden' => test_zci($jan_9, html => $jan_9_html),
```
Comment (jagtalon) [inline t/NameDays.t]: @mintsoft some of the tests are failing on our machines such as this one. Do you think it needs something like Mocktime?

**Comment by jagtalon** [comment]
@W25 @abeyang https://ddh1.duckduckgo.com/?q=1+June+name+day&ia=answer

File: share/goodie/name_days/Bulgaria.txt [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+
+Hristofor
+Ksenia
+Grigor, Grigorena
+Zhivka, Zhivko, Zoya
+
+
+	
+
+
```
Comment (jagtalon) [inline share/goodie/name_days/Bulgaria.txt]: Why are there a lot of whitespace?

File: lib/DDG/Goodie/NameDays.pm [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+description "Name Days for a given name or date";
+primary_example_queries "name day Maria", "1 June name day";
+secondary_example_queries "name days today", "imieniny 9 stycznia", "imieniny Marii";
+category "dates";
+topics "social", "everyday";
+code_url "https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/NameDays/W25/W25.pm";
+attribution github => ["http://github.com/W25", "W25"];
+
+# Triggers
+triggers any => "name day", "name days", "nameday", "namedays", "imieniny", "jmeniny", "svátek";
```
Comment (jagtalon) [inline lib/DDG/Goodie/NameDays.pm]: Could you add a comment what "imieniny", "jmeniny", and "svátek" are?

File: lib/DDG/Goodie/NameDays.pm [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        if (!$day) {
+            $day = parse_other_date_formats($_);
+        }
+
+        return unless $day;
+
+        # Any leap year here, because the array includes February, 29
+        $day->set_year(2000);
+        
+        $query = time2str( '%B %o', $day->epoch() );
```
Comment (jagtalon) [inline lib/DDG/Goodie/NameDays.pm]: Is it possible that this will fail?

**Comment by jagtalon** [comment]
@W25 Looks like the same thing is happening here: https://github.com/duckduckgo/zeroclickinfo-goodies/pull/921#issuecomment-70606954 (double `zci__body`).

File: share/goodie/name_days/Bulgaria.txt (hunk 2 above)
Comment (W25) [inline share/goodie/name_days/Bulgaria.txt]: The file should contain 366 lines (one for each day of a year). There are no name days for these days in Wikipedia.

Comment (jagtalon) [inline share/goodie/name_days/Bulgaria.txt]: @W25 Oh I understand. Thanks

**Comment by jagtalon** [comment]
@W25 

- Why does it show May 31 when I search June 1? Maybe some timezone thing?

screen shot 2015-01-21 at 11 56 40 am

- For June 1, why are there only 9 countries but in http://www.namedaycalendar.com/june/1 there are 16?

**Comment by abeyang** [comment]
Hi, in addition to what @jagtalon said, can we have these css tweaks? They're mostly all padding fixes... give the text some more breathing room:

```css
.name-days-country-name {
  padding-left: 0.4em; 
}

.name-days-country {
/* REMOVE ALL OTHER PADDING INSIDE THIS CLASS! */
  padding-right: 2em;
}

.name-days-country-name, .name-days-tile {
  line-height: 1.6;
}

/* should add a class to <tr> in html and add it here */
tr td {
  padding-top: 0.3em;
  padding-bottom: 0.3em;
}
```

Finally, can we change the tab ("Name Day" instead of "Answer") and title to look more like this (applies to both date and name):

name-days-v2 3-maria

Quite an easy fix -- just need to change it to an h3 with this class:

screen shot 2015-01-21 at 9 33 36 am

Whew, quite a lot -- hope it's not too overwhelming. Thanks again for working on this!

**Comment by jagtalon** [comment]
@abeyang @W25 we currently can't change the tab name for a Goodie, and "Show More" is a bit difficult to pull off.

@abeyang thank you!

**Comment by abeyang** [comment]
@W25 given @jagtalon 's recent comment (that we're unable to change tab name), let's keep the title the way you have it, at least for dates:

> Name Days on (Date)

For searches based on names, can we do this:

> Days Named (Name)

Where "(Name)" is Maria, etc. Hope that makes sense...

**Comment by moollaza** [comment]
ping! @W25 just checking if you're still interested in finishing this one up? We're very close to launching it live!

My biggest concern is the relevancy issues @jagtalon pointed out above: https://github.com/duckduckgo/zeroclickinfo-goodies/pull/743#issuecomment-70875432 -- any ideas whats going on there?

If you're able to address that I'll gladly finish up the design stuff and merge this in so we can launch it live!

**Comment by W25** [comment]
@jagtalon @moollaza  I can only guess why it fails on the server. Here is a screenshot from DuckPan server; it works okay here:

1june

**Comment by W25** [comment]
@jagtalon @moollaza  Sorry, but I can only guess why it fails on the server. The same query is okay on a DuckPan server. I replaced time2str with own implementation, please try again if it works for you.

**Comment by moollaza** [comment]
>I found data only for 11 countries in Wikipedia. Namedaycalendar.com is not open-source, so we cannot use it.

@W25 understood, maybe we can get the community to source out more namedays for other countries.

@jagtalon can you put this on a server for testing?

**Comment by jagtalon** [comment]
It's up in https://ddh9.duckduckgo.com/?q=name+day+Maria&ia=answer!

**Comment by moollaza** [comment]
@W25 this is looking awesome!

**Comment by abeyang** [comment]
It's getting there, @W25. Couple more things: please make the change that @jagtalon [proposed here](https://github.com/duckduckgo/zeroclickinfo-goodies/pull/743#issuecomment-70693983) (removing the double zci__body): 

Second, can we add these css changes? Basically, making the color darker and the columns the same size:

```css
.zci__body, .zero_click_answer {
  color: #333;
}

td.name-days-country {
  width: 25%;
}

td.name-days-dates {
  width: 75%;
}
```

Thanks!

**Comment by W25** [comment]
@abeyang Done.

**Comment by abeyang** [comment]
Whoa @W25 , so sorry -- I must've missed this update. Looks good! :+1: /cc @moollaza

**Comment by moollaza** [comment]
@W25 thanks so much for the tremendous work. I can't tell you how much we appreciate you sticking around to get through this with us. I'm really happy to see how great the final product us.

Feel free to stick around and comment/help on other PR's & Issues or even submit more Instant Answers! It's my goal to make sure PR's don't take this long to get merged and now that the Goodie templates are basically done (seriously, they are!) it should be much smoother moving forward for Goodie devs to produce awesome looking IAs.

**Comment by moollaza** [comment]
@W25 congrats! It's live: https://duckduckgo.com/?q=name+day+Maria&ia=answer

Thanks again for taking the time to contribute! We really appreciate it. Feel free to stick around and comment/help on other PR's & Issues or even submit more Instant Answers!

**Pull request merged**